### ISSUE TITLE
Optifine assumes the server port is always 25565 on Java 9+

### ISSUE THREAD

**Issue by SecretlyJealous** [sp614x/optifine#2917]
Using vanilla Minecraft 1.14.4 with Java 13
--
In this example I'll be connecting to:

![Untitled](https://user-images.githubusercontent.com/10485646/65708739-3079b480-e087-11e9-95fe-d81e5f8d7b58.png)

I am able to ping the server:

![Untitled](https://user-images.githubusercontent.com/10485646/65708000-86e5f380-e085-11e9-8aea-f588cec6d087.png)

Here's what the logs show:
```
[17:52:29] [Client thread/INFO]: Connecting to phan024.ddns.net., 25565
[17:52:29] [Client thread/WARN]: Unknown custom packed identifier: minecraft:register
[17:52:32] [Client thread/WARN]: Unknown custom packed identifier: minecraft:register
[17:52:32] [Client thread/INFO]: [CHAT]                                      
[17:52:32] [Client thread/INFO]: Loaded 7 advancements
```

Everything works correctly when using default Minecraft but there are problems when I start using Optifine.

OptiFine 1.14.4 HD U F4 pre5 with Java 13
--
I am connecting to the same server however I am unable to ping the server or connect to it:

![Untitled](https://user-images.githubusercontent.com/10485646/65707586-c233f280-e084-11e9-9053-75a2c081b467.png)

Here's what the logs show:
```
[17:42:13] [main/INFO]: Connecting to play.phanaticmc.com, 25565
[17:42:13] [Server Connector #1/ERROR]: Couldn't connect to server
java.net.SocketException: Network is unreachable
```

The issue is that Optifine with Java 9 and above for some reason will use the IP address you have specified and will disregard any SRV records the server has set in place. The log shows that when I'm using Optifine with Java 13 it is trying to connect me to `play.phanaticmc.com, 25565` however when using default Minecraft 1.14.4 with Java 13 it is pinging the server correctly and allows me to connect, the log shows that it is taking me to `phan024.ddns.net., 25565`. For some reason Optifine is not doing DNS lookups while regular Minecraft does do DNS lookups. I'm aware that Optifine does not intefere with network related things, but I believe this issue IS related to Optifine because it only happens when I'm using Optifine and the logs prove that Optifine is not behaving the same as regular Minecraft when attempting to connect to servers that use SRV records.

**Comment by Puremin0rez** [comment]
I can definitely confirm this issue.

I am unable to connect to servers that primarily use SRV records using Optifine 1.14.4 and Java 11.

**Comment by sp614x** [comment]
OptiFine doesn't change anything related to multiplayer, network or server communication.
The problem comes from Java itself and is probably linked to signed/unsigned JARs and the way that OptiFine is loaded via Launchwrapper. Only Java 9 and newer versions are affected.

**Comment by sp614x** [comment]
The DNS SRV resolving is done via reflection and Java 9 restricts reflection calls between modules.
Vanilla is running from the classpath while OptiFine is being loaded via LaunchWrapper. Apparently Java 9 sets different reflection restrictions depending from where the code is loaded.
The exception is:
```
javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.dns.DnsContextFactory
	at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:719) ~[?:?]
	at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305) ~[?:?]
	at javax.naming.InitialContext.init(InitialContext.java:236) ~[?:?]
	at javax.naming.InitialContext.<init>(InitialContext.java:208) ~[?:?]
	at javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101) ~[?:?]
	at dkh.b(ServerAddress.java:107) ~[dkh.class:?]
	at dkh.a(ServerAddress.java:80) ~[dkh.class:?]
	at dkk.a(SourceFile:52) ~[dkk.class:?]
	at der$d.d(SourceFile:233) ~[der$d.class:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:299) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) [?:?]
	at java.lang.Thread.run(Thread.java:844) [?:?]
Caused by: java.lang.ClassNotFoundException: com.sun.jndi.dns.DnsContextFactory
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:185) ~[launchwrapper-of-2.1.jar:2.1]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:563) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:496) ~[?:?]
	at java.lang.Class.forName0(Native Method) ~[?:?]
	at java.lang.Class.forName(Class.java:375) ~[?:?]
	at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:107) ~[?:?]
	at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:82) ~[?:?]
	at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:716) ~[?:?]
	... 14 more
```

**Comment by sp614x** [comment]
With Launchwrapper bugfix to show the real problem:
```
[OptiFine] getServerAddress
javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.dns.DnsContextFactory [Root exception is java.lang.IllegalAccessException: class javax.naming.spi.NamingManager (in module java.naming) cannot access class com.sun.jndi.dns.DnsContextFactory (in module jdk.naming.dns) because module jdk.naming.dns does not export com.sun.jndi.dns to module java.naming]
	at java.naming/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:719)
	at java.naming/javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
	at java.naming/javax.naming.InitialContext.init(InitialContext.java:236)
	at java.naming/javax.naming.InitialContext.<init>(InitialContext.java:208)
	at java.naming/javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:130)
	at dkh.b(ServerAddress.java:107)
	at dkh.a(ServerAddress.java:80)
	at dkk.a(SourceFile:52)
	at der$d.d(SourceFile:233)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: java.lang.IllegalAccessException: class javax.naming.spi.NamingManager (in module java.naming) cannot access class com.sun.jndi.dns.DnsContextFactory (in module jdk.naming.dns) because module jdk.naming.dns does not export com.sun.jndi.dns to module java.naming
	at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:376)
	at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:639)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:490)
	at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:166)
	at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:404)
	at java.base/java.lang.Class.newInstance(Class.java:590)
	at java.naming/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:716)
	... 14 more
```

**Comment by EternalMC** [comment]
So can we fix it?

**Comment by Jawastew** [comment]
I'm sorry this isn't adding much to the conversation.
I just spent the last 2+ hours trying to get my SRV records to work and finally decided to try it with a fresh install/different machine. Which narrowed it down to optifine.
The only thing I kept observing was Minecraft adamantly checking A and AAAA records but not SRV records. Which just didn't give the right search results :(
I'm really glad I finally found this Issue thread, as it gives me some closure :)

Using OptiFine_1.14.4_HD_U_F4, Java 10.0.2 on Ubuntu

**Comment by wpalmer** [comment]
"me too", OptiFine_1.14.4_HD_U_F4, OpenJDK 11.0.4 2019-07-16, Ubuntu 18.04.3 LTS

that is to say: this also happens using OpenJDK.

However, this does *not* seem to be an issue when using Optifine+Fabric+MultiMC
ie: OptiFine_1.14.4_HD_U_F4, optifabric-0.6.0, MultiMC 0.6.7-1290, OpenJDK 11.0.4, Ubuntu 18.04.3 LTS

**Comment by sp614x** [comment]
This only happens when OptiFine is installed standalone.
Loading by Forge or Fabric works differently and has no problems.

**Comment by FireController1847** [comment]
I can confirm this is still a bug in OptiFine 1.15.2 HD U G1 pre10

**Comment by Vladinuto** [comment]
Still a bug on OptiFine 1.15.2 HD U G1 pre13

**Comment by sslanec** [comment]
Still happening in pre15. Here's a copy of what I wrote [here](https://github.com/sp614x/optifine/issues/2579#issuecomment-623212259) before finding this thread:

I'm having this issue with the latest preview (1.15 pre15). I can connect to minecraft.rtgame.co.uk fine with the vanilla client on Java 8 or 11, but only with Java 8 using Optifine. Otherwise it just sits at "connecting..." until it times out. Even weirder, I can connect fine using Forge+Optifine on Java 11, but you can't leave spawn with a non-vanilla client. Hypixel and Mineplex work fine with either version of Java, so I'm not sure if it's a problem with this specific server or what.

**Comment by sslanec** [comment]
Still happening in OptiFine 1.15.2 HD U G1 pre15

**Comment by outsidefactor** [comment]
I can confirm that resolving SRV records still fails in G pre16 on Linux.

It doesn't impact vanilla and using a numerical address and port prevents this from happening.

As it seems to only impact Optifine Standalone it suggests that's where the issue lies. It doesn't impact Vanilla, and it doesn't impact Forge installs. It is the same for both the included Java exe and the latest OpenJDK. Has anyone had this issue with the latest Oracle JRE/JDK for Linux?

What is different about Optifine? Does Optifine invoke the Java network stack differently than Vanilla or Forge?

**Comment by Qwertylex** [comment]
same issue on 1.15.2 HD U G1 pre18

**Comment by Strum355** [comment]
This is a known issue with Minecraft in general. There was a bug report to Mojang about this and theyre not fixing it as of yet. This issue should be closed

**Comment by SecretlyJealous** [comment]
> This is a known issue with Minecraft in general. There was a bug report to Mojang about this and theyre not fixing it as of yet. Use a lower Java version. This issue should be closed

This is not a known issue with Minecraft, please link the ticket if so. You can use any java version with Minecraft, Optifine creates the problem.

**Comment by Strum355** [comment]
https://bugs.mojang.com/browse/MC-175531

**Comment by SecretlyJealous** [comment]
That is a separate issue related to Java on Linux not Windows

**Comment by Puremin0rez** [comment]
> 
> 
> This is a known issue with Minecraft in general. There was a bug report to Mojang about this and theyre not fixing it as of yet. Use a lower Java version. This issue should be closed

This is a known issue with Optifine, and you'd see that if you actually spent time testing the issue or [reading the comments by sp641x himself](https://github.com/sp614x/optifine/issues/2917#issuecomment-555195641) above confirming it's an issue with Optifine.

This is happening on both Windows and Linux (and presumably Mac).

Vanilla + Java 9+ = Proper SRV Resolution
Optifine + Java 9+ = Broken SRV Resolution
Fabric + Optifine (Optifabric) + Java 9+ = Proper SRV Resolution
Optifine 1.15 + Java 8 = Proper SRV Resolution

While the ticket you linked may indeed be an actual issue on Linux (or the reporter was actually using Optifine, that report is pretty vague), this is happening on all platforms with Optifine and has already been confirmed. Please use due diligence when requesting for issues to be closed that you haven't even tested in the future :)

**Comment by ZekerZhayard** [comment]
https://github.com/sp614x/LegacyLauncher/pull/1

**Comment by smartin1018** [comment]
This is not fixed in the latest version

**Comment by SecretlyJealous** [comment]
@sp614x Would you be willing to merge https://github.com/sp614x/LegacyLauncher/pull/1 and see if it fixes the issue?

**Comment by sp614x** [comment]
Fixed in 1.16.2 pre4 with the bugfix https://github.com/sp614x/LegacyLauncher/pull/1
